File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/visiting/StdDevEvaluator.java
package org.opencds.cqf.cql.engine.elm.visiting;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;
import org.opencds.cqf.cql.engine.execution.State;
import org.opencds.cqf.cql.engine.runtime.Quantity;

import java.math.BigDecimal;
import java.util.List;

/*
StdDev(argument List<Decimal>) Decimal
StdDev(argument List<Quantity>) Quantity

The StdDev operator returns the statistical standard deviation of the elements in source.
If the source contains no non-null elements, null is returned.
If the list is null, the result is null.
Return types: BigDecimal & Quantity
*/

public class StdDevEvaluator {

    public static Object stdDev(Object source, State state) {

        if (source == null) {
            return null;
        }

        if (source instanceof Iterable) {

            if (((List<?>) source).isEmpty()) {
                return null;
            }

            Object variance = VarianceEvaluator.variance(source, state);

            return variance instanceof BigDecimal ?
                    PowerEvaluator.power(variance, new BigDecimal("0.5")) :
                    new Quantity().withValue((BigDecimal) PowerEvaluator.power(((Quantity) variance).getValue(),
                            new BigDecimal("0.5"))).withUnit(((Quantity) variance).getUnit());
        }

        throw new InvalidOperatorArgument(
                "StdDev(List<Decimal>) or StdDev(List<Quantity>)",
                String.format("StdDev(%s)", source.getClass().getName())
        );
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/StdDevEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.cqframework.cql.shared.BigDecimal
import org.opencds.cqf.cql.engine.execution.State
import org.opencds.cqf.cql.engine.runtime.Decimal
import org.opencds.cqf.cql.engine.runtime.Quantity
import org.opencds.cqf.cql.engine.runtime.Value
import org.opencds.cqf.cql.engine.runtime.toCqlDecimal

/*
StdDev(argument List<Decimal>) Decimal
StdDev(argument List<Quantity>) Quantity

The StdDev operator returns the statistical standard deviation of the elements in source.
If the source contains no non-null elements, null is returned.
If the list is null, the result is null.
Return types: BigDecimal & Quantity
*/
object StdDevEvaluator {
    @JvmStatic
    fun stdDev(source: Value?, state: State?): Value? {
        if (source == null) {
            return null
        }
        return when (val variance = VarianceEvaluator.variance(source, state, true, "StdDev")) {
            // The cases in which Variance returns null are the same as those where StdDev does.
            null -> return null
            is Decimal -> PowerEvaluator.power(variance, BigDecimal("0.5").toCqlDecimal())
            else -> {
                // If variance is a Quantity, we made sure that the unit part was not squared during
                // the variance computation. As a result, we can take the square root of the value
                // but keep the unit as it is.
                val value =
                    PowerEvaluator.power(
                        (variance as Quantity).value?.toCqlDecimal(),
                        BigDecimal("0.5").toCqlDecimal(),
                    ) as Decimal
                Quantity().withValue(value.value).withUnit(variance.unit)
            }
        }
    }
}
